db/cache/charts.go

```diff
@@ -913,82 +931,93 @@ func blockTimes(blocks ChartUints) (ChartUints, ChartUints) {
+			avgDiffs = append(avgDiffs, pts.Avg(0, len(pts)))
 			nextIdx++
 			if nextIdx > len(ticks)-1 {
 				break
 			}
-			workingOn = next
 			lastIdx = i
 			next = ticks[nextIdx]
 		}
 	}
-	return times, avgs
+	return avgDiffs
 }
 
-func blockSizeChart(charts *ChartData, zoom ZoomLevel) ([]byte, error) {
-	switch zoom {
-	case BlockZoom:
+func blockSizeChart(charts *ChartData, bin binLevel, axis axisType) ([]byte, error) {
+	if bin == DayBin && axis == HeightAxis {
+		return charts.encode(charts.Days.Height, charts.Days.BlockSize)
+	}
+	switch bin {
+	case BlockBin:
 		return charts.encode(charts.Blocks.Time, charts.Blocks.BlockSize)
-	case DayZoom:
+	case DayBin:
 		return charts.encode(charts.Days.Time, charts.Days.BlockSize)
 	}
-	return nil, InvalidZoomErr
+	return nil, InvalidBinErr
 }
 
-func blockchainSizeChart(charts *ChartData, zoom ZoomLevel) ([]byte, error) {
-	switch zoom {
-	case BlockZoom:
+func blockchainSizeChart(charts *ChartData, bin binLevel, axis axisType) ([]byte, error) {
+	if bin == DayBin && axis == HeightAxis {
+		return charts.encode(charts.Days.Height, accumulate(charts.Days.BlockSize))
+	}
+	switch bin {
+	case BlockBin:
 		return charts.encode(charts.Blocks.Time, accumulate(charts.Blocks.BlockSize))
-	case DayZoom:
+	case DayBin:
 		return charts.encode(charts.Days.Time, accumulate(charts.Days.BlockSize))
 	}
-	return nil, InvalidZoomErr
+	return nil, InvalidBinErr
 }
 
-func chainWorkChart(charts *ChartData, zoom ZoomLevel) ([]byte, error) {
-	switch zoom {
-	case BlockZoom:
+func chainWorkChart(charts *ChartData, bin binLevel, axis axisType) ([]byte, error) {
+	if bin == DayBin && axis == HeightAxis {
+		return charts.encode(charts.Days.Height, charts.Days.Chainwork)
+	}
+	switch bin {
+	case BlockBin:
 		return charts.encode(charts.Blocks.Time, charts.Blocks.Chainwork)
-	case DayZoom:
+	case DayBin:
 		return charts.encode(charts.Days.Time, charts.Days.Chainwork)
 	}
-	return nil, InvalidZoomErr
+	return nil, InvalidBinErr
 }
 
-func coinSupplyChart(charts *ChartData, zoom ZoomLevel) ([]byte, error) {
-	switch zoom {
-	case BlockZoom:
+func coinSupplyChart(charts *ChartData, bin binLevel, axis axisType) ([]byte, error) {
+	if bin == DayBin && axis == HeightAxis {
+		return charts.encode(charts.Days.Height, accumulate(charts.Days.NewAtoms), charts.Days.Height)
+	}
+	switch bin {
+	case BlockBin:
 		return charts.encode(charts.Blocks.Time, accumulate(charts.Blocks.NewAtoms))
-	case DayZoom:
+	case DayBin:
 		return charts.encode(charts.Days.Time, accumulate(charts.Days.NewAtoms), charts.Days.Height)
 	}
-	return nil, InvalidZoomErr
+	return nil, InvalidBinErr
 }
 
-func durationBTWChart(charts *ChartData, zoom ZoomLevel) ([]byte, error) {
-	switch zoom {
-	case BlockZoom:
+func durationBTWChart(charts *ChartData, bin binLevel, axis axisType) ([]byte, error) {
+	if bin == DayBin && axis == HeightAxis {
+		return charts.encode(charts.Days.Height, avgBlockTimes(charts.Days.Time, charts.Blocks.Time))
```

Comment:
But `avgBlockTimes` function does that validation.

Reply:
`avgBlockTimes` returns a slice of length `len(charts.Days.Time) - 1`, but `charts.Days.Height` is length `len(charts.Days.Height)` = `len(charts.Days.Time)`, so it has one more element. 

Now that I look at it again, even though you are encoding slices of different lengths, since `Height` is the longer one, it will be truncated during encoding by removing an element from the end. So by luck, you will end up with the correct axis alignment. But it's much better IMO to just create the data sets correctly to begin with instead of relying on a fallback feature of `encode`. 

So actually, I had it backwards, the correct way to size the x axis is `charts.Days.Height[:len(charts.Days.Height)-1]`, since we associate a day with it's midnight.

Reply:
I thought returning to your previous implementation was better as I would not have to check for the array length.

Reply:
The previous implementation did not deal with `Height` arrays. You're just adding those now, so you have to adjust the height array. It just needs the last element removed.